Feature 10861

ashish-greycube opened the pull request:
Please refer design flow chart as below.
![image](https://user-images.githubusercontent.com/29812965/31996803-725f4a30-b9a7-11e7-943f-7ad0b3a54469.png)

Animated gif explaining the "Bypass Customer Credit Limit" check feature
and "Customer Credit Balance Report" in action.

Outstanding Amount is not updated on submission of "Sales Order".
It is updated during full/partial payment against "Sales Order" and/or after submitting "Delivery Note". (as per the case)
![image](https://user-images.githubusercontent.com/29812965/31996875-b2223128-b9a7-11e7-957f-a17833ad526e.png)

Test Case :
Automated test cases passed for both scenarios.

   1. With Credit Limit check Bypass at Sales Order 
    
![image](https://user-images.githubusercontent.com/29812965/31997070-4fc2e59e-b9a8-11e7-9443-cda7d75366d5.png)

2. With Credit Limit check without bypass at Sales Order (default)
![image](https://user-images.githubusercontent.com/29812965/31997097-65f52c00-b9a8-11e7-920f-4f5af53bb9ff.png)

ashish-greycube committed:
customer.json -- a checkbox field is added which gives user ability to bypass credit limit check at sales order level for a customer. test_customer.js : auto generated file as customer.json is modified

ashish-greycube committed:
sales_order.py -- Three changes are done (a) if bypass credit limit check is enabled we should not check customer credit on submit of sales order (b.1/b.2) There is provision to make delivery note and sales invoice from sales order. Since credit limit is bypassed at sales order level we need to check credit of customer on make of (b.1)delivery note or (b.2)sales invoice

ashish-greycube committed:
sales_invoice.py -- Check credit limit when make delivery is done from sales invoice.We need to check the credit limit as  there could be multiple partially paid sales order(note we are bypassing credit check at sales order) so we need to recheck credit balance of customer to avoid any delivery crossing credit limit from sales invoice

ashish-greycube committed:
customer_credit_balance.py -- Credit limit should now reflect outstanding amount for each customer based on value of flage bypass credit limit check at sales order level. A new column is added to show the value of this flag. customer.py -- get_credit_limit function which gets called by the above customer_credit_balance report is updated to deduct the amount of sales order for which credit limit bypass flag is set

ashish-greycube committed:
New test case files added.(a)test_sales_order_with_bypass_credit_limit_check.js -- New feature: i.e. bypass credit limit check at sales order.(b)test_sales_order_without_bypass_credit_limit_check.js --Regression test case : the existing default flow is checked i.e. credit limit check happens at sales order level

ashish-greycube commented:
This code check is in context to https://github.com/frappe/erpnext/issues/10861

tundebabzy commented on erpnext/selling/doctype/sales_order/sales_order.py:
```diff
@@ -196,7 +196,11 @@ def update_project(self):
-		check_credit_limit(self.customer, self.company)
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# bypass credit limit check is set to true (1) at sales order level, then we need not to check credit limit and vise versa
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", self.customer, "bypass_credit_limit_check_at_sales_order")
```
bypass_credit_limit_check_at_sales_order = cint(frappe.db.........)

tundebabzy commented on erpnext/selling/doctype/sales_order/sales_order.py:
```diff
@@ -461,6 +465,14 @@ def set_missing_values(source, target):
+	    
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at delivery note
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order")
```
bypass_credit_limit_check_at_sales_order = cint(frappe.db....)

tundebabzy commented on erpnext/selling/doctype/sales_order/sales_order.py:
```diff
@@ -524,6 +536,13 @@ def set_missing_values(source, target):
 
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at sales invoice
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order")
```
bypass_credit_limit_check_at_sales_order = cint(frappe.db.....)

tundebabzy commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -70,7 +70,6 @@ def validate(self):
 		self.clear_unallocated_advances("Sales Invoice Advance", "advances")
 		self.add_remarks()
 		self.validate_write_off_account()
-		self.validate_duplicate_offline_pos_entry()
```
Why did you remove?

tundebabzy commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -463,12 +462,6 @@ def validate_write_off_account(self):
 		if flt(self.write_off_amount) and not self.write_off_account:
 			msgprint(_("Please enter Write Off Account"), raise_exception=1)
 
-	def validate_duplicate_offline_pos_entry(self):
```
Why did you remove?

tundebabzy commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -924,6 +917,13 @@ def set_missing_values(source, target):
 
+	    #PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+		# Since the credit limit check is bypassed at sales order level, we need to check it at delivery note
+		bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order")
```
bypass_credit_limit_check_at_sales_order = cint(frappe.db....)

tundebabzy commented on erpnext/selling/doctype/customer/customer.py:
```diff
@@ -186,6 +188,20 @@ def get_customer_outstanding(customer, company):
+	#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+	# Since the credit limit check is bypassed at sales order level, when customer credit balance report is run we need to treat sales order with status 'To Deliver and Bill' as not outstanding
+	outstanding_based_on_bypassed_so = 0.0
+	bypass_credit_limit_check_at_sales_order = frappe.db.get_value("Customer", customer, "bypass_credit_limit_check_at_sales_order")
```
bypass_credit_limit_check_at_sales_order = cint(frappe.db....)

tundebabzy commented on erpnext/selling/doctype/customer/test_customer.js:
```diff
@@ -0,0 +1,23 @@
+/* eslint-disable */
```
This looks like the default template. How about update it or remove it completely.

tundebabzy commented on erpnext/selling/doctype/customer/customer.json:
```diff
@@ -911,6 +911,36 @@
+   "search_index": 0, 
+   "set_only_once": 0, 
+   "unique": 0
+  }, 
```
Explicitly set default value for `bypass_credit_limit_check_at_sales_order` to 0 since explicit is always better.

tundebabzy commented:
- After setting default to 0, you will also need to add a patch to update old records.
- Still missing python tests
- Still missing updates to the manual about this new feature

tundebabzy commented on erpnext/selling/report/customer_credit_balance/customer_credit_balance.py:
```diff
@@ -49,6 +52,7 @@ def get_details(filters):
-	return frappe.db.sql("""select name, customer_name from `tabCustomer` %s""" 
+	#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+	# Return column bypass credit limit check at sales order level	
+	return frappe.db.sql("""select name, customer_name,bypass_credit_limit_check_at_sales_order from `tabCustomer` %s""" 
```
This would cause an SQL Injection vulnerability. Please correct this

ashish-greycube committed:
customer.json -- a checkbox field is added which gives user ability to bypass credit limit check at sales order level for a customer. It is default to 0. Also patch is added to update value of new field to 0 after migrate.
test_customer.js -- It is auto generated file as customer.json is modified. And it is removed.

ashish-greycube committed:
sales_order.py -- Three changes are done (a) if bypass credit limit check is enabled we should not check customer credit on submit of sales order (b.1/b.2) There is provision to make delivery note and sales invoice from sales order. Since credit limit is bypassed at sales order level we need to check credit of customer on make of (b.1)delivery note or (b.2)sales invoice. cint function is added.

ashish-greycube committed:
sales_invoice.py -- Check credit limit when make delivery is done from sales invoice.We need to check the credit limit as  there could be multiple partially paid sales order(note we are bypassing credit check at sales order) so we need to recheck credit balance of customer to avoid any delivery crossing credit limit from sales invoice.cint function is added.

ashish-greycube committed:
customer.py -- get_credit_limit function which gets called by the above customer_credit_balance report is updated to deduct the amount of sales order for which credit limit bypass flag is set. cint function is added.

ashish-greycube commented:
Hi @tundebabzy , please find the update as requested
1)All the comments are incorporated
2)Still missing updates to the manual about this new feature. Ashish: Documentation changes pull request is [11343](https://github.com/frappe/erpnext/pull/11343)
3)customer_credit_balance.py :This would cause an SQL Injection vulnerability. Please correct this. Ashish: I have just added a new column to select query. The structure of SQL is as it is and not changed by me.
4)Still missing python tests. Ashish : JS test case are written which covers the new flow. Also regression test case is added. This are created using [guide](https://frappe.io/docs/user/en/guides/automated-testing/qunit-testing) . Test result are also shared. Test case are self sufficient i.e. It creates the requierd test data.

ashish-greycube commented:
by mistake i closed it

ashish-greycube commented:
@tundebabzy please let me know if it is good to go!

ashish-greycube commented:
@nabinhait , It appears the pull is assigned to you. Please guide. thanks

ashish-greycube commented:
hi @nabinhait  @tundebabzy , I have meet all the requirements or changes suggested. Please guide. Also note that in this case javascript test case will only make sense as when submit button is clicked from UI the new business logic code gets executed, whereas python test case of sales order submits directly.
So, please review and help me to merge my first pull request. thanks!

ashish-greycube commented:
Hi @nabinhait request you to help me merge the PR. thanks!

nabinhait commented on erpnext/patches/v9_0/set_bypasscreditlimitcheckatsalesorder_in_customer.py:
```diff
@@ -0,0 +1,10 @@
+	if "bypass_credit_limit_check_at_sales_order" in frappe.db.get_table_columns("Customer"):
+		frappe.db.sql("""
+			update `tabCustomer`
+			set bypass_credit_limit_check_at_sales_order = 0
```
I think the patch is not needed, system should set "0" by default.

nabinhait commented on erpnext/selling/doctype/customer/customer.py:
```diff
@@ -186,6 +188,20 @@ def get_customer_outstanding(customer, company):
+	outstanding_based_on_bypassed_so = 0.0
+	bypass_credit_limit_check_at_sales_order = cint(frappe.db.get_value("Customer", customer, "bypass_credit_limit_check_at_sales_order"))
+	if bypass_credit_limit_check_at_sales_order == 1 and caller=='customer_credit_balance_report':
+		outstanding_based_on_bypassed_so = frappe.db.sql("""
```
Why this query needed?
We can skip `outstanding_based_on_so` optionally, like in case of report.

ashish-greycube replied:
Nabin, there are two things (1) outstanding_based_on_so and (2) outstanding_based_on_bypassed_so [where credit limit is overruled].
When customer_credit_balance_report is run we should return correct outstanding amount by checking each customer for flag "by pass customer credit limit".
We need to treat sales order with status 'To Deliver and Bill'  and where no advance is paid as NOT outstanding
Hence the logic is "outstanding for by pass credit limit" = "outstanding_based_on_so" ---- (minus) "outstanding_based_on_bypassed_so"

nabinhait commented on erpnext/selling/doctype/customer/customer.py:
```diff
@@ -170,7 +170,9 @@ def check_credit_limit(customer, company):
-def get_customer_outstanding(customer, company):
+#PR : 10861, Author : ashish-greycube & jigneshpshah,  Email:mr.ashish.shah@gmail.com 
+# get_customer_outstanding is a very generic function which is invoked from many places. A third non mandatory argument is added to change its behaviour based on caller .
+def get_customer_outstanding(customer, company,caller=None):
```
The report name should not be hard-coded like this using caller. Tomorrow, if we need to call this function for another purpose, will not be able to do that. It should be a flag, based on purpose.

nabinhait replied:
Always use space after comma

nabinhait commented:
Closing it for now, please reopen after fixing.

ashish-greycube commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -924,6 +924,13 @@ def set_missing_values(source, target):
+		bypass_credit_limit_check_at_sales_order = cint(frappe.db.get_value("Customer", source.customer, "bypass_credit_limit_check_at_sales_order"))
+		if bypass_credit_limit_check_at_sales_order == 1:
+			from erpnext.selling.doctype.customer.customer import check_credit_limit
+			check_credit_limit(source.customer, source.company)
```
Agreed.It is removed

ashish-greycube commented on erpnext/patches.txt:
```diff
@@ -455,3 +455,4 @@ erpnext.patches.v9_0.add_healthcare_domain
 erpnext.patches.v9_0.set_variant_item_description
 erpnext.patches.v9_0.set_uoms_in_variant_field
 erpnext.patches.v9_0.copy_old_fees_field_data
+erpnext.patches.v9_0.set_bypasscreditlimitcheckatsalesorder_in_customer
```
As the patch is removed. File is no longer required
